tests/test-awful-rules.lua

```diff
@@ -282,10 +282,38 @@ test_rule {
 
         assert(c.screen.selected_tag.name ~= "8")
 
+        -- Test the custom sources
+        assert(c.random1)
+        assert(c.random2)
+        assert(not c.random3)
+
         return true
     end
 }
 
+-- Test the custom sources
+assert(awful.rules.add_rule_source("before", function(c, props, _)
+    assert(type(c) == "client")
+    assert(not props.random1)
+
+    props.random2 = true
+end, {"awful.rules"}, {"awful.spawn"}))
+
+assert(awful.rules.add_rule_source("before2", function(c, props, _)
+    assert(type(c) == "client")
+    props.random3 = true
+    assert(false)
+end, {"awful.rules"}, {"awful.spawn"}))
+
+assert(awful.rules.remove_rule_source("before2"))
+
+assert(awful.rules.add_rule_source("after", function(c, props, _)
+    assert(type(c) == "client")
+    assert(props.random2)
+
+    props.random1 = true
+end, {"awful.spawn", "void"}))
+
```

Comment:
Could you move the `test_rule` call to down here? I was going to complain that `random1` is unused, but instead it is checked by code that appears earlier in this file...